File: experiments/adrianco/experiment-1/runs/language=typescript_model=sonnet_tooling=beads/rep2/src/__tests__/books.test.ts
import request from 'supertest';
import { createApp } from '../app';
import { createTestDb } from '../database';
import Database from 'better-sqlite3';

describe('Books API', () => {
  let db: Database.Database;
  let app: ReturnType<typeof createApp>;

  beforeEach(() => {
    db = createTestDb();
    app = createApp(db);
  });

  afterEach(() => {
    db.close();
  });

  // --- Health check ---
  describe('GET /health', () => {
    it('returns 200 with status ok', async () => {
      const res = await request(app).get('/health');
      expect(res.status).toBe(200);
      expect(res.body).toEqual({ status: 'ok' });
    });
  });

  // --- POST /books ---
  describe('POST /books', () => {
    it('creates a book with valid data', async () => {
      const res = await request(app)
        .post('/books')
        .send({ title: 'Clean Code', author: 'Robert Martin', year: 2008, isbn: '9780132350884' });

      expect(res.status).toBe(201);
      expect(res.body.id).toBeDefined();
      expect(res.body.title).toBe('Clean Code');
      expect(res.body.author).toBe('Robert Martin');
      expect(res.body.year).toBe(2008);
      expect(res.body.isbn).toBe('9780132350884');
    });

    it('creates a book with only required fields', async () => {
      const res = await request(app)
        .post('/books')
        .send({ title: 'Minimal Book', author: 'Some Author' });

      expect(res.status).toBe(201);
      expect(res.body.title).toBe('Minimal Book');
      expect(res.body.author).toBe('Some Author');
      expect(res.body.year).toBeNull();
      expect(res.body.isbn).toBeNull();
    });

    it('returns 400 when title is missing', async () => {
      const res = await request(app)
        .post('/books')
        .send({ author: 'Some Author' });

      expect(res.status).toBe(400);
      expect(res.body.error).toMatch(/title/i);
    });

    it('returns 400 when author is missing', async () => {
      const res = await request(app)
        .post('/books')
        .send({ title: 'Some Title' });

      expect(res.status).toBe(400);
      expect(res.body.error).toMatch(/author/i);
    });

    it('returns 400 when title is empty string', async () => {
      const res = await request(app)
        .post('/books')
        .send({ title: '   ', author: 'Author' });

      expect(res.status).toBe(400);
    });
  });

  // --- GET /books ---
  describe('GET /books', () => {
    beforeEach(async () => {
      await request(app).post('/books').send({ title: 'Book A', author: 'Alice' });
      await request(app).post('/books').send({ title: 'Book B', author: 'Bob' });
      await request(app).post('/books').send({ title: 'Book C', author: 'Alice' });
    });

    it('returns all books', async () => {
      const res = await request(app).get('/books');
      expect(res.status).toBe(200);
      expect(res.body).toHaveLength(3);
    });

    it('filters books by author', async () => {
      const res = await request(app).get('/books?author=Alice');
      expect(res.status).toBe(200);
      expect(res.body).toHaveLength(2);
      expect(res.body.every((b: { author: string }) => b.author === 'Alice')).toBe(true);
    });

    it('returns empty array when no books match filter', async () => {
      const res = await request(app).get('/books?author=Nobody');
      expect(res.status).toBe(200);
      expect(res.body).toHaveLength(0);
    });
  });

  // --- GET /books/:id ---
  describe('GET /books/:id', () => {
    it('returns a specific book', async () => {
      const created = await request(app)
        .post('/books')
        .send({ title: 'Target Book', author: 'Author' });

      const res = await request(app).get(`/books/${created.body.id}`);
      expect(res.status).toBe(200);
      expect(res.body.title).toBe('Target Book');
    });

    it('returns 404 for non-existent book', async () => {
      const res = await request(app).get('/books/99999');
      expect(res.status).toBe(404);
    });

    it('returns 400 for invalid id', async () => {
      const res = await request(app).get('/books/not-a-number');
      expect(res.status).toBe(400);
    });
  });

  // --- PUT /books/:id ---
  describe('PUT /books/:id', () => {
    it('updates a book', async () => {
      const created = await request(app)
        .post('/books')
        .send({ title: 'Old Title', author: 'Old Author', year: 2000 });

      const res = await request(app)
        .put(`/books/${created.body.id}`)
        .send({ title: 'New Title', author: 'New Author', year: 2024 });

      expect(res.status).toBe(200);
      expect(res.body.title).toBe('New Title');
      expect(res.body.author).toBe('New Author');
      expect(res.body.year).toBe(2024);
    });

    it('returns 404 for non-existent book', async () => {
      const res = await request(app)
        .put('/books/99999')
        .send({ title: 'Title', author: 'Author' });

      expect(res.status).toBe(404);
    });

    it('returns 400 if title is cleared', async () => {
      const created = await request(app)
        .post('/books')
        .send({ title: 'Title', author: 'Author' });

      const res = await request(app)
        .put(`/books/${created.body.id}`)
        .send({ title: '', author: 'Author' });

      expect(res.status).toBe(400);
    });
  });

  // --- DELETE /books/:id ---
  describe('DELETE /books/:id', () => {
    it('deletes a book', async () => {
      const created = await request(app)
        .post('/books')
        .send({ title: 'To Delete', author: 'Author' });

      const delRes = await request(app).delete(`/books/${created.body.id}`);
      expect(delRes.status).toBe(204);

      const getRes = await request(app).get(`/books/${created.body.id}`);
      expect(getRes.status).toBe(404);
    });

    it('returns 404 when deleting non-existent book', async () => {
      const res = await request(app).delete('/books/99999');
      expect(res.status).toBe(404);
    });
  });
});

File: experiments/adrianco/experiment-38-alllang-80b-fullctx/bookshop/runs/agent=hermes-local_language=elixir_model=mlxlocal/mlx-community--Qwen3-Coder-Next-4bit_prompt=neutral_stack=m80/rep3/book_api/test/book_api/books_test.exs
defmodule BookAPI.BooksTest do
  use ExUnit.Case, async: true

  alias BookAPI.Books

  describe "list_books/1" do
    test "returns all books" do
      book = insert(:book, title: "Test Book", author: "Test Author")
      books = Books.list_books()
      assert length(books) == 1
      assert hd(books).title == "Test Book"
    end

    test "filters books by author" do
      insert(:book, title: "Book 1", author: "Author A")
      insert(:book, title: "Book 2", author: "Author B")
      insert(:book, title: "Book 3", author: "Author A")

      books = Books.list_books(author: "Author A")
      assert length(books) == 2
      Enum.each(books, fn b -> assert b.author == "Author A" end)
    end
  end

  describe "get_book!/1" do
    test "returns the book with given id" do
      book = insert(:book, title: "Test Book", author: "Test Author")
      retrieved_book = Books.get_book!(book.id)
      assert retrieved_book.title == "Test Book"
    end

    test "raises if book not found" do
      assert_raise Ecto.NoResultsError, fn ->
        Books.get_book!(999_999)
      end
    end
  end

  describe "create_book/1" do
    test "creates a book with valid data" do
      attrs = %{title: "New Book", author: "New Author", year: 2024, isbn: "1234567890"}
      {:ok, %Book{} = book} = Books.create_book(attrs)
      assert book.title == "New Book"
      assert book.author == "New Author"
      assert book.year == 2024
      assert book.isbn == "1234567890"
    end

    test "returns error with invalid data (missing required fields)" do
      attrs = %{year: 2024, isbn: "1234567890"}
      {:error, %Ecto.Changeset{}} = Books.create_book(attrs)
    end

    test "validates title is required" do
      attrs = %{author: "Some Author", year: 2024}
      {:error, %Ecto.Changeset{errors: errors}} = Books.create_book(attrs)
      assert {:title, {"can't be blank", [_]}} in errors
    end

    test "validates author is required" do
      attrs = %{title: "Some Title", year: 2024}
      {:error, %Ecto.Changeset{errors: errors}} = Books.create_book(attrs)
      assert {:author, {"can't be blank", [_]}} in errors
    end
  end

  describe "update_book/2" do
    test "updates a book with valid data" do
      book = insert(:book, title: "Old Title", author: "Old Author")
      attrs = %{title: "New Title", author: "New Author"}
      {:ok, %Book{} = updated_book} = Books.update_book(book, attrs)
      assert updated_book.title == "New Title"
      assert updated_book.author == "New Author"
    end

    test "returns error with invalid data" do
      book = insert(:book, title: "Old Title", author: "Old Author")
      attrs = %{year: 2024}
      {:error, %Ecto.Changeset{}} = Books.update_book(book, attrs)
    end
  end

  describe "delete_book/1" do
    test "deletes a book" do
      book = insert(:book, title: "To Delete", author: "Delete Author")
      {:ok, %Book{} = deleted_book} = Books.delete_book(book)
      assert deleted_book.title == "To Delete"

      assert Books.list_books() == []
    end
  end

  defp insert(:book, attrs \\ %{}) do
    {:ok, book} =
      attrs
      |> Map.put_new(:title, "Test Book")
      |> Map.put_new(:author, "Test Author")
      |> BookAPI.Books.create_book()

    book
  end
end
